compiler/src/diagnostics/lenses.re

```diff
@@ -2,76 +2,154 @@ open Grain_typed;
-  let lensList: lens_list_t = {lenses: lenses};
-  let json_list = lens_list_t_to_yojson(lensList);
-  Yojson.Basic.pretty_to_string(Yojson.Safe.to_basic(json_list));
+let make_value =
+    (~location: Grain_parsing.Location.t, ~sigStr: string, ~valType: string) => {
+  let (file, startline, startchar, sbol) =
+    get_raw_pos_info(location.loc_start);
+  let (_, endline, endchar, ebol) = get_raw_pos_info(location.loc_end);
+  let lens: lens_t = {
+    sl: startline,
+    sc: startchar,
+    sb: sbol,
+    el: endline,
+    ec: endchar,
+    eb: ebol,
+    s: sigStr,
+    t: valType,
+  };
+  lens;
 };
 
-let output_lenses = (program: Typedtree.typed_program): list(lens_t) => {
-  let lenses =
-    List.map(
-      (statement: Typedtree.toplevel_stmt) => {
-        let d = statement.ttop_desc;
-        switch (d) {
-        | TTopExpr(expr) =>
-          let buf = Buffer.create(64);
-          let ppf = Format.formatter_of_buffer(buf);
-          let t: Types.type_expr = expr.exp_type;
-          Printtyp.type_expr(ppf, t);
-          Format.pp_print_flush(ppf, ());
-          let signt = Buffer.contents(buf);
-          let lens: lens_t = {
-            line: statement.ttop_loc.loc_start.pos_lnum,
-            signature: signt,
-          };
-          Some(lens);
-        | TTopLet(_, _, _, valueBindings) =>
-          let signatures =
-            List.map(
-              (vb: Typedtree.value_binding) => {
-                let expr = vb.vb_expr;
-
-                let t: Types.type_expr = expr.exp_type;
-                let buf = Buffer.create(64);
-                let ppf = Format.formatter_of_buffer(buf);
-                Printtyp.type_expr(ppf, t);
-                Format.pp_print_flush(ppf, ());
-                Buffer.contents(buf);
-              },
-              valueBindings,
-            );
-          if (List.length(signatures) > 0) {
-            let signature = List.hd(signatures);
-            let lens: lens_t = {
-              line: statement.ttop_loc.loc_start.pos_lnum,
-              signature,
-            };
-            Some(lens);
-          } else {
-            None;
-          };
+let hover_val =
+    (
+      ~location: Grain_parsing.Location.t,
+      ~t: Types.type_expr,
+      ~valType: string,
+    ) => {
+  let buf = Buffer.create(64);
+  let ppf = Format.formatter_of_buffer(buf);
+  Printtyp.type_expr(ppf, t);
+  Format.pp_print_flush(ppf, ());
+  let sigStr = Buffer.contents(buf);
+  make_value(~location, ~valType, ~sigStr);
+};
 
-        | _ => None
-        };
+let create_record_signature = (labels: list(Types.record_field)) => {
+  let recSig =
+    List.fold_left(
+      (acc, field: Types.record_field) => {
+        let comma = acc == "" ? "" : ", ";
+        let buf = Buffer.create(64);
+        let ppf = Format.formatter_of_buffer(buf);
+        Printtyp.type_expr(ppf, field.rf_type);
+        Format.pp_print_flush(ppf, ());
+        acc
+        ++ comma
```

Comment:
Maybe it might be better to use `Printf.sprintf` here?

Reply:
Would love to!  Can you give me a code sample?  This pretty printer stuff confuses the heck of out me!